specification/storagecache/resource-manager/Microsoft.StorageCache/preview/2020-08-13-aml-preview/storagecache.json

```diff
@@ -0,0 +1,2623 @@
+          }
+        },
+        "x-ms-long-running-operation": true,
+        "x-ms-long-running-operation-options": {
+          "final-state-via": "azure-async-operation"
+        },
+        "tags": [
+          "Caches"
+        ],
+        "description": "Upgrade a Cache's firmware if a new version is available. Otherwise, this operation has no effect.",
+        "operationId": "Caches_UpgradeFirmware"
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}/providers/Microsoft.StorageCache/amlFilesystems": {
+      "get": {
+        "produces": [
+          "application/json"
+        ],
+        "parameters": [
+          {
+            "description": "Target resource group.",
+            "in": "path",
+            "name": "resourceGroupName",
+            "required": true,
+            "type": "string"
+          },
+          {
+            "$ref": "#/parameters/ApiVersionParameter"
+          },
+          {
+            "$ref": "#/parameters/SubscriptionIdParameter"
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "A list of amlFilesystem objects. Note that entity references might replace complete amlFilesystem objects, as described in http://docs.oasis-open.org/odata/odata-json-format/v4.01/cs01/odata-json-format-v4.01-cs01.html#sec_EntityReference",
+            "schema": {
+              "$ref": "#/definitions/amlFilesystemsListResult"
+            }
+          },
+          "default": {
+            "description": "Error response describing why the operation failed.",
+            "schema": {
+              "$ref": "#/definitions/CloudError"
+            }
+          }
+        },
+        "x-ms-examples": {
+          "amlFilesystems_ListByResourceGroup": {
+            "$ref": "./examples/amlFilesystems_ListByResourceGroup.json"
+          }
+        },
+        "x-ms-pageable": {
+          "nextLinkName": "nextLink"
+        },
+        "tags": [
+          "amlFilesystems"
+        ],
+        "description": "Returns all amlFilesystems the user has access to under a resource group.",
+        "operationId": "amlFilesystems_ListByResourceGroup"
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}/providers/Microsoft.StorageCache/amlFilesystems/{amlFilesystemName}": {
+      "delete": {
+        "produces": [
+          "application/json"
+        ],
+        "parameters": [
+          {
+            "description": "Target resource group.",
+            "in": "path",
+            "name": "resourceGroupName",
+            "required": true,
+            "type": "string"
+          },
+          {
+            "description": "Name of amlFilesystem. Length of name must be not greater than 80 and chars must be in list of [-0-9a-zA-Z_] char class.",
+            "in": "path",
+            "name": "amlFilesystemName",
+            "pattern": "^[-0-9a-zA-Z_]{1,80}$",
```

Comment:
Copied this from Cache.
Dan, any reasoning on the start with -?
I can change this, just want to check if there was a rule I missed.

This is based on "$ref": "#/definitions/ResourceName"
"ResourceName": {
      "description": "Schema for the name of resources served by this provider. Note that objects will contain an odata @id annotation as appropriate. This will contain the complete URL of the object. These names are case-preserving, but not case sensitive.",
      "pattern": "^[-0-9a-zA-Z_]{1,80}$",
      "type": "string"
    },

Reply:
The starting - means that names can contain dashes.   But that pattern allows a leading dash which is probably not kosher.    
 A one character name probably isn't kosher either, is the lower limit for ARM 3 chars?  I forget.

I'd suggest:

"pattern": "^[0-9a-zA-Z_][-0-9a-zA-Z_]{2,79}$"

Reply:
Here's the naming rules for various resources: https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/resource-name-rules

In this case, I suggest a slightly modified version of Dan's:
"pattern": "^[0-9a-zA-Z][-0-9a-zA-Z_]{0,78}[0-9a-zA-Z]$"

The logical description of that is:
Length: 2-80
Valid characters: Alphanumerics, underscores, and hyphens. Start and end with alphanumeric.

Reply:
FWIW, I'm explicitly trying to forbid names like '___', '_-_', etc.

Reply:
fixed, see on next commit.